### PR TITLE
✨ Add entry for SLSA Go builder

### PR DESCRIPTION
**Pull request by laurentsimon**
## Pre-requisites

- [ ] Prior to submitting a new workflow, please apply to join the GitHub Technology Partner Program: [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner).

---

### **Please note that at this time we are only accepting new starter workflows for Code Scanning. Updates to existing starter workflows are fine.**

---

## Tasks

**For _all_ workflows, the workflow:**

- [x] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [ ] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [x] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [x] Should include comments in the workflow for any parts that are not obvious or could use clarification.
- [x] Should specify least priviledge [permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) for `GITHUB_TOKEN` so that the workflow runs successfully.

**For _CI_ workflows, the workflow:**

- [x] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [ ] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [ ] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [x] Packaging workflows should run on `release` with `types: [ created ]`.
- [x] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**For _Code Scanning_ workflows, the workflow:**

- [ ] Should be preserved under [the `code-scanning` directory](https://github.com/actions/starter-workflows/tree/main/code-scanning).
- [x] Should include a matching `code-scanning/properties/*.properties.json` file (for example, [`code-scanning/properties/codeql.properties.json`](https://github.com/actions/starter-workflows/blob/main/code-scanning/properties/codeql.properties.json)), with properties set as follows:
  - [ ] `name`: Name of the Code Scanning integration.
  - [ ] `organization`: Name of the organization producing the Code Scanning integration.
  - [ ] `description`: Short description of the Code Scanning integration.
  - [ ] `categories`: Array of languages supported by the Code Scanning integration.
  - [ ] `iconName`: Name of the SVG logo representing the Code Scanning integration. This SVG logo must be present in [the `icons` directory](https://github.com/actions/starter-workflows/tree/main/icons).
- [ ] Should run on `push` to `branches: [ $default-branch, $protected-branches ]` and `pull_request` to `branches: [ $default-branch ]`. We also recommend a `schedule` trigger of `cron: $cron-weekly` (for example, [`codeql.yml`](https://github.com/actions/starter-workflows/blob/c59b62dee0eae1f9f368b7011cf05c2fc42cf084/code-scanning/codeql.yml#L14-L21)).

**Some general notes:**

- [ ] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [ ] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [ ] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [ ] Automation and CI workflows cannot be dependent on a paid service or product.

### CODE REVIEW

**Comment by laurentsimon** [comment]
Note: this is the GA for https://github.blog/2022-04-07-slsa-3-compliance-with-github-actions/
/cc @josepalafox

/cc @MarkLodato @joshuagl @ianlewis @asraa

@MarkLodato: we need an icon. Which shall we use?

**Comment by laurentsimon** [comment]
talked offline with @asraa. We've picked the generic SLSA logo from the website

File: ci/properties/go-publisher-ossf-slsa3.properties.json
Comment (laurentsimon) [reply ci/properties/go-publisher-ossf-slsa3.properties.json:5]: thanks. Done.

**Comment by laurentsimon** [comment]
> Don't you need a pure icon, not one with text? If so, then https://github.com/slsa-framework/slsa/blob/main/resources/assets/logo/SLSA-logo.svg would be appropriate instead.

Updated the PR with this logo. Thanks

**Comment by laurentsimon** [comment]
friendly ping. Are there any blockers to merge this PR?
We're waiting for the merge to finalize the blog post https://github.com/slsa-framework/slsa/pull/398#pullrequestreview-1008849289

File: ci/properties/go-publisher-ossf-slsa3.properties.json
Comment (bishal-pdMSFT) [inline ci/properties/go-publisher-ossf-slsa3.properties.json:6]: Add "Go" as well

Comment (laurentsimon) [reply ci/properties/go-publisher-ossf-slsa3.properties.json:6]: I'll defer to @josepalafox

I think we suggested using `SLSA verified builder` instead, since CI is a general term and this workflow is about secure build pipeline, namely SLSA

Wdut?

File: ci/go-publisher-ossf-slsa3.yml
Comment (bishal-pdMSFT) [inline ci/go-publisher-ossf-slsa3.yml:1]: Please add a header disclaimer that this workflow uses third party actions. See the [guidance](https://github.com/actions/starter-workflows/blob/main/.github/pull_request_template.md).
```
# This workflow uses actions that are not certified by GitHub.
# They are provided by a third-party and are governed by
# separate terms of service, privacy policy, and support
# documentation.
```

Comment (bishal-pdMSFT) [inline ci/go-publisher-ossf-slsa3.yml:1]: Since this workflow also publishes the built asset, please put a suffix of `-publish`. See [guidance](https://github.com/actions/starter-workflows/blob/main/.github/pull_request_template.md).

May be `go-ossf-slsa3-publish.yml`

File: ci/go-publisher-ossf-slsa3.yml
Comment (bishal-pdMSFT) [inline ci/go-publisher-ossf-slsa3.yml:17]: Please use full 40 character sha ref for hardened dependency

File: ci/go-publisher-ossf-slsa3.yml
Comment (bishal-pdMSFT) [inline ci/go-publisher-ossf-slsa3.yml:1]: Please add a short summary:
1. What does this starter workflow light up for customer?
2. Any links for "more details", "support"

**Comment by bishal-pdMSFT** [comment]
I am not an expert in svg, but the icon does not look like a svg file. Please check. It also does not get rendered in the GitHub editor.

**Review by bishal-pdMSFT — CHANGES_REQUESTED** [review]
Left some comments

File: ci/go-publisher-ossf-slsa3.yml
Comment (laurentsimon) [reply ci/go-publisher-ossf-slsa3.yml:17]: Re-usable workflows are not be on-par with standard actions when using the OIDC token (https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token).

The workflow branch is not reported in the OIDC token for the re-usable workflow: it just has `wrkflow_path@ref`

This is a current limitation of the GH token, and we've made a feature request to address this with @josepalafox

Practically, this means that the hash is not guaranteed to be from the main branch, which means it's worse to pin by hash: any of the maintainers can push a different branch without code reviews, and users would be using it and it would pass the verification of the [slsa-framework/slsa-verifier](https://github.com/slsa-framework/slsa-verifier)

Until GH adds support, we've decided to pin by full version `v1.2.3` instead.

Hope you understand.

File: ci/go-publisher-ossf-slsa3.yml
Comment (laurentsimon) [reply ci/go-publisher-ossf-slsa3.yml:1]: Done

Comment (laurentsimon) [reply ci/go-publisher-ossf-slsa3.yml:1]: Done.

Comment (laurentsimon) [reply ci/go-publisher-ossf-slsa3.yml]: Done. PTAL

File: ci/go-publisher-ossf-slsa3.yml
Comment (ianlewis) [reply ci/go-publisher-ossf-slsa3.yml]: This probably deserves a good comment noting why the hash isn't used in this case.

Comment (laurentsimon) [reply ci/go-publisher-ossf-slsa3.yml]: Any suggestion? Trying to keep it simple for users, but not sure how much we need say.

Comment (ianlewis) [reply ci/go-publisher-ossf-slsa3.yml]: Ah, I see this gets included as a template in their repo? Then it makes sense to leave it off.

File: ci/go-publisher-ossf-slsa3.yml
Comment (bishal-pdMSFT) [reply ci/go-publisher-ossf-slsa3.yml]: Third party is in sense that it is not directly owned by GitHub.

File: ci/go-publisher-ossf-slsa3.yml
Comment (bishal-pdMSFT) [reply ci/go-publisher-ossf-slsa3.yml]: Got it. Thanks for the explanation. I agree to leave it off in case of reusable workflows.

File: ci/go-publisher-ossf-slsa3.yml
Comment (bishal-pdMSFT) [reply ci/go-publisher-ossf-slsa3.yml]: I don't see this updated in the PR. Did you may be miss to push the change?

Comment (bishal-pdMSFT) [reply ci/go-publisher-ossf-slsa3.yml]: Unresolving the comment as I do not see this updated in file.

File: ci/properties/go-publisher-ossf-slsa3.properties.json
Comment (bishal-pdMSFT) [reply ci/properties/go-publisher-ossf-slsa3.properties.json]: Unfortuantely we do not have a category for "secure build" and hence we will use CI category. We will internally discuss this and plan for adding more categories but that will take time.

File: ci/properties/go-ossf-slsa3-publish.properties.json
Comment (bishal-pdMSFT) [inline ci/properties/go-ossf-slsa3-publish.properties.json:5]: update the icon name here. PR check is failing.

File: ci/go-publisher-ossf-slsa3.yml
Comment (laurentsimon) [reply ci/go-publisher-ossf-slsa3.yml]: my bad, updated now. Note that I don't fully understand the motivation behind this disclaimer. An Action can be replaced by an internal library or piece of code and achieve the same functionality, without the need for the disclaimer.

I would understand it better if it was required for 3P actions external to this repo .

File: ci/properties/go-ossf-slsa3-publish.properties.json
Comment (laurentsimon) [reply ci/properties/go-ossf-slsa3-publish.properties.json]: done, sorry I forgot to push.

File: ci/go-publisher-ossf-slsa3.yml
Comment (laurentsimon) [reply ci/go-publisher-ossf-slsa3.yml]: done, PTAL

File: ci/properties/go-ossf-slsa3-publish.properties.json
Comment (bishal-pdMSFT) [inline ci/properties/go-ossf-slsa3-publish.properties.json:4]: Too long description, will clutter the UI. Revert back to older one -> `Compile your Golang project using a SLSA3 compliant builder.`
Other details have been added at top of yml file so not needed here.

File: ci/go-ossf-slsa3-publish.yml
Comment (bishal-pdMSFT) [inline ci/go-ossf-slsa3-publish.yml:30]: Not a comment for this PR but something to think about for future in case you plan to add more workflows for building other languages. Can they be encapsulated in a single workflow rather than a separate one for each language?
For example, look at [codeql](https://github.com/actions/starter-workflows/blob/874ee449ed7c7203faec42e7708b2929d123ff2b/code-scanning/properties/codeql.properties.json#L6) one which caters to multiple languages.

**Review by bishal-pdMSFT — APPROVED** [review]
LGTM apart from a small comment about description property.

File: ci/go-ossf-slsa3-publish.yml
Comment (laurentsimon) [reply ci/go-ossf-slsa3-publish.yml:30]: Thanks for the tip. I was initially thinking of having different workflows for each language. Since most projects contain a single "main" project, it would avoid noise. I'll think about this for future languages, thanks!

File: ci/properties/go-ossf-slsa3-publish.properties.json
Comment (laurentsimon) [reply ci/properties/go-ossf-slsa3-publish.properties.json]: Done, thank you

**Comment by laurentsimon** [comment]
I've addressed all the comments. Ready to merge on my side. Thank you so much for you comments and patience!

**Review by bishal-pdMSFT — APPROVED** [review]

**Pull request merged**
